src/ui/public/vis/map/kibana_map.js

```diff
@@ -620,13 +624,38 @@ export class KibanaMap extends EventEmitter {
     }
   }
 
-  _getTMSBaseLayer(options) {
-    return L.tileLayer(options.url, {
-      minZoom: options.minZoom,
-      maxZoom: options.maxZoom,
-      subdomains: options.subdomains || []
-    });
-  }
+  _getTMSBaseLayer = (function () {
+    let tilesCreated = 0;
+    let tilesLoaded = 0;
+    let tilesErrored = 0;
+    let tmsBaseLayer;
+    const checkIsValid = () => {
+      const loadingComplete = tilesCreated === tilesLoaded + tilesErrored;
+      if (loadingComplete) {
+        // If any tiles have loaded, is valid
+        tilesLoaded ? null : defaultLayerLoadWarning();
+      }
+    };
+    const { createTile } = L.TileLayer.prototype;
+
+    return ({ url, minZoom, maxZoom, subdomains }) => {
+      tmsBaseLayer = new (L.TileLayer.extend({
+        createTile: function (coords, done) {
+          tilesCreated++;
+          return createTile.call(this, coords, done);
+        }
+      }))(url, {
+        minZoom: minZoom,
+        maxZoom: maxZoom,
+        subdomains: subdomains || []
+      });
+      tmsBaseLayer.on('tileload', () => tilesLoaded++);
+      tmsBaseLayer.on('tileerror', () => tilesErrored++);
+      tmsBaseLayer.on('tileload tileerror', checkIsValid);
+
+      return tmsBaseLayer;
+    };
+  }());
```

Comment:
Putting all of this in a IIFE makes the logic very confusing and hard to follow. Think you could refactor this a little so it makes more sense?